### PR TITLE
add toDigest(from hex string to bytes) lazy

### PR DESCRIPTION
**Pull request by Tynukua**
alternative for https://github.com/dlang/phobos/pull/8523

### CODE REVIEW

**Comment by Tynukua** [comment]
@atilaneves @thewilsonator @RazvanN7 
I created a new PR I'm not really sure what is better...

File: std/digest/package.d [hunk 1]
```diff
@@ -1212,3 +1212,141 @@ if (isInputRange!R1 && isInputRange!R2 && !isInfinite!R1 && !isInfinite!R2 &&
...
+    import std.ascii : isHexDigit;
+
+    private string hex;
+    private size_t byteCount;
+    private size_t hexPos;
+    private size_t bytesPos;
+    private bool addZero;
+    private char[2] _front;
+
+    @safe @nogc pure nothrow
```
Comment (RazvanN7) [inline std/digest/package.d:1283]: You could just use: `@safe @nogc pure nothrow:` . That will get rid of a level of indentation.

File: std/digest/package.d [hunk 2]
```diff
@@ -1212,3 +1212,141 @@ if (isInputRange!R1 && isInputRange!R2 && !isInfinite!R1 && !isInfinite!R2 &&
...
+        this(string t)
+        {
+            hex = t;
+            if (hex.length >= 2 && hex[0 .. 2] == "0x")
+            {
+                hex = hex[2 .. $];
+            }
+            foreach (c; hex)
+            {
+                byteCount += c.isHexDigit;
```
Comment (RazvanN7) [inline std/digest/package.d:1294]: Iterating through all the digits upfront is wasteful. You could do the `c.isHexDigit` in `front`/`popFront`.

Comment (RazvanN7) [reply std/digest/package.d:1294]: Also, I don't think the `byteCount` field is required. You could always check `c.length`

File: std/digest/package.d [hunk 3]
```diff
@@ -1212,3 +1212,141 @@ if (isInputRange!R1 && isInputRange!R2 && !isInfinite!R1 && !isInfinite!R2 &&
...
+
+    testToStaticDigest!3([255, 0, 0], "0xff");
+    testToStaticDigest!3([0, 255, 0], "0x00ff");
+    testToStaticDigest!1([0x12], "0x12ff");
+    testToStaticDigest!2([0x12, 255], "0x12ff");
+    testToStaticDigest!2([0x12, 255], "0x12__ff");
+    testToStaticDigest!4([0x1, 0x23, 0xaa, 0xaa], "0x123AaGGAA");
+}
+
+private struct HexToBytesConverter
```
Comment (RazvanN7) [inline std/digest/package.d:1272]: This struct should be declared as `static` inside the `toDigest` function. There's no point in polluting the namespace.

File: std/digest/package.d (hunk 2 above)
Comment (Tynukua) [reply std/digest/package.d:1294]: > Iterating through all the digits upfront is wasteful. You could do the `c.isHexDigit` in `front`/`popFront`.

how to proceed with such strings like "0x1aa"
so if we don't count all bytes we could add bytes only to end like [0x1a, 0xa0];

but IMO when we see such a string we wanna parse it like [0x01, 0xaa]

Comment (Tynukua) [reply std/digest/package.d:1294]: > Also, I don't think the `byteCount` field is required. You could always check `c.length`

I wanna filter non-hex values like '_'
f.e. to parse "0xffaa_ffaa"

File: std/digest/package.d (hunk 3 above)
Comment (Tynukua) [reply std/digest/package.d:1272]: > This struct should be declared as `static` inside the `toDigest` function. There's no point in polluting the namespace.

but its already private...

**Comment by aG0aep6G** [comment]
Please don't call it `toDigest`. The function doesn't do any digesting. It does the opposite of the existing `toHexString`, so call it `fromHexString`.

**Pull request closed without merging**